mink/tasks/constraints_task.py

```diff
@@ -0,0 +1,145 @@
+        
+        # Get the mask for equality constraints
+        self._constraints_mask = np.argwhere(
+            mj_data.efc_type == mujoco.mjtConstraint.mjCNSTR_EQUALITY
+        ).ravel()
+        
+        # Update the number of constraints
+        self._n_constraints = len(self._constraints_mask)
+        
+        # Update the cost vector if needed
+        if self._n_constraints > 0:
+            if len(self.constraint_cost) == 1:
+                # Use the same cost for all constraints
+                self.cost = np.full(self._n_constraints, self.constraint_cost[0])
+            elif len(self.constraint_cost) == self._n_constraints:
+                # Use the provided cost vector
+                self.cost = self.constraint_cost
+            else:
+                raise TaskDefinitionError(
+                    f"{self.__class__.__name__} constraint cost should be a scalar or "
+                    f"a vector of length {self._n_constraints} but got "
+                    f"{len(self.constraint_cost)}"
+                )
+
+    def compute_error(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the constraint error.
+
+        The error is extracted directly from the MuJoCo data structure, specifically
+        from the `efc_pos` field which contains the constraint violations.
+
+        Args:
+            configuration: Robot configuration :math:`q`.
+
+        Returns:
+            Constraint error vector.
+        """
+        # Update constraints information
+        self._update_constraints_info(configuration)
+        
+        # If there are no constraints of the specified type, return an empty array
+        if self._n_constraints == 0:
+            return np.zeros(1)
+        
+        # Extract the constraint error from MuJoCo data
+        mj_data = configuration.data
+        error = mj_data.efc_pos.copy()[self._constraints_mask]
+        
+        return error
+
+    def compute_jacobian(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the constraint Jacobian.
+
+        The Jacobian is extracted directly from the MuJoCo data structure, specifically
+        from the `efc_J` field which contains the constraint Jacobian. For equality
+        constraints, this Jacobian relates changes in the configuration to changes in
+        the constraint violations:
+
+        .. math::
+
+            \dot{c}(q) = J(q) \dot{q}
+
+        where :math:`c(q)` is the constraint function and :math:`J(q)` is its Jacobian.
+
+        Args:
+            configuration: Robot configuration :math:`q`.
+
+        Returns:
+            Constraint Jacobian matrix :math:`J(q)`.
+        """
+        # Update constraints information if needed
+        if self._constraints_mask is None:
+            self._update_constraints_info(configuration)
+        
+        # If there are no constraints of the specified type, return an empty array
+        if self._n_constraints == 0:
```

Comment:
So having the cost dynamically change in size is kinda scaring me. If we ensure we always initialize to a keyframe where the equality constraint(s) are satisfied, then we can get rid of the special case where n_constraints=0. What are your thoughts? Obviously your approach is more robust. 

But if you do `configuration.update_from_keyframe("home")` in the main script, then you'll notice you can get rid of these if conditions and the whole thing runs fine.